include/beast/websocket/impl/write.ipp

```diff
@@ -244,101 +168,335 @@ operator()(error_code ec, bool again)
+            if(d.ws.rd_op_.maybe_invoke())
+            {
+                d.state = do_maybe_suspend;
+                d.ws.get_io_service().post(
+                    std::move(*this));
                 return;
             }
-            auto const n = clamp(d.remain, d.tmp_size);
-            mutable_buffers_1 mb{d.tmp, n};
-            buffer_copy(mb, d.cb);
-            d.cb.consume(n);
+            d.state = d.entry;
+            break;
+
+        //----------------------------------------------------------------------
+
+        case do_mask_nofrag:
+        {
+            d.remain = buffer_size(d.cb);
+            d.fh.fin = d.fin;
+            d.fh.len = d.remain;
+            d.fh.key = d.ws.maskgen_();
+            detail::prepare_key(d.key, d.fh.key);
+            detail::write<static_streambuf>(
+                d.fh_buf, d.fh);
+            auto const n =
+                clamp(d.remain, d.ws.wr_.buf_size);
+            auto const b =
+                buffer(d.ws.wr_.buf.get(), n);
+            buffer_copy(b, d.cb);
+            detail::mask_inplace(b, d.key);
             d.remain -= n;
-            detail::mask_inplace(mb, d.key);
-            // send header and payload
-            d.state = d.remain > 0 ? 2 : 99;
+            // Send frame header and partial payload
+            d.state = d.remain == 0 ?
+                do_upcall : do_mask_nofrag + 1;
             BOOST_ASSERT(! d.ws.wr_block_);
             d.ws.wr_block_ = &d;
             boost::asio::async_write(d.ws.stream_,
-                buffer_cat(d.fh_buf.data(),
-                    mb), std::move(*this));
+                buffer_cat(d.fh_buf.data(), b),
+                    std::move(*this));
             return;
         }
 
-        // sent masked payload
-        case 2:
+        case do_mask_nofrag + 1:
         {
-            auto const n = clamp(d.remain, d.tmp_size);
-            mutable_buffers_1 mb{d.tmp,
-                static_cast<std::size_t>(n)};
-            buffer_copy(mb, d.cb);
-            d.cb.consume(n);
+            d.cb.consume(d.ws.wr_.buf_size);
+            auto const n =
+                clamp(d.remain, d.ws.wr_.buf_size);
+            auto const b =
+                buffer(d.ws.wr_.buf.get(), n);
+            buffer_copy(b, d.cb);
+            detail::mask_inplace(b, d.key);
             d.remain -= n;
-            detail::mask_inplace(mb, d.key);
-            // send payload
+            // Send parial payload
             if(d.remain == 0)
-                d.state = 99;
-            BOOST_ASSERT(d.ws.wr_block_ == &d);
+                d.state = do_upcall;
             boost::asio::async_write(
-                d.ws.stream_, mb, std::move(*this));
+                d.ws.stream_, b, std::move(*this));
+            return;
+        }
+
+        //----------------------------------------------------------------------
+
+        case do_mask_frag:
+        {
+            auto const n = clamp(
+                d.remain, d.ws.wr_.buf_size);
```

Comment:
Why bother `clamp`ing if you're jumping to `do_mast_nofrag` when `d.remain <= d.ws.wr_.buf_size`?